pkg/aws/eni/node.go

```diff
@@ -336,26 +335,43 @@ func (n *Node) CreateInterface(ctx context.Context, allocation *ipam.AllocationA
 		return 0, errUnableToDetermineLimits, fmt.Errorf(errUnableToDetermineLimits)
 	}
 
-	bestSubnet := n.manager.FindSubnetByTags(n.k8sObj.Spec.ENI.VpcID, n.k8sObj.Spec.ENI.AvailabilityZone, n.k8sObj.Spec.ENI.SubnetTags)
+	n.mutex.RLock()
+	resource := *n.k8sObj
+	n.mutex.RUnlock()
+
+	bestSubnet := n.manager.FindSubnetByTags(resource.Spec.ENI.VpcID, resource.Spec.ENI.AvailabilityZone, resource.Spec.ENI.SubnetTags)
 	if bestSubnet == nil {
-		return 0, errUnableToFindSubnet, fmt.Errorf("No matching subnet available for interface creation (VPC=%s AZ=%s SubnetTags=%s",
-			n.k8sObj.Spec.ENI.VpcID, n.k8sObj.Spec.ENI.AvailabilityZone, n.k8sObj.Spec.ENI.SubnetTags)
+		return 0,
+			errUnableToFindSubnet,
+			fmt.Errorf(
+				"No matching subnet available for interface creation (VPC=%s AZ=%s SubnetTags=%s",
+				resource.Spec.ENI.VpcID,
+				resource.Spec.ENI.AvailabilityZone,
+				resource.Spec.ENI.SubnetTags,
+			)
 	}
 
-	securityGroupIDs, err := n.getSecurityGroupIDs(ctx)
+	securityGroupIDs, err := n.getSecurityGroupIDs(ctx, resource.Spec.ENI)
 	if err != nil {
-		return 0, errUnableToGetSecurityGroups, fmt.Errorf("%s %s", errUnableToGetSecurityGroups, err)
+		return 0,
+			errUnableToGetSecurityGroups,
+			fmt.Errorf("%s %s", errUnableToGetSecurityGroups, err)
 	}
 
-	desc := "Cilium-CNI (" + n.node.InstanceID() + ")"
+	n.mutex.RLock()
+	instanceID := n.node.InstanceID()
+	n.mutex.RUnlock()
```

Comment:
I've double checked, the `n.node` does not need to be protected by the mutex. We can drop this change and also modify the location of `n.mutex` to be below `node *ipam.Node`, on lines 46-49, to avoid confusion.